### PR TITLE
Integration of `nn.aggr` within `MessagPassing`

### PR DESCRIPTION
**Pull request by lightaime**
Support new aggregation modules with `MessagePassing`.

### CODE REVIEW

File: torch_geometric/nn/aggr/basic.py [hunk 1]
```diff
@@ -22,7 +22,8 @@ def forward(self, x: Tensor, index: Optional[Tensor] = None, *,
         return self.reduce(x, index, ptr, dim_size, dim, reduce='sum')
 
 
-AddAggregation = SumAggregation  # Alias
+AddAggregation = type('AddAggregation', SumAggregation.__bases__,
```
Comment (rusty1s) [inline torch_geometric/nn/aggr/basic.py:25]: Mmh, I see the problem. Let us instead add `"add"` support directly in `aggregation_resolver`. This should be cleaner.

File: torch_geometric/nn/conv/message_passing.py [hunk 2]
```diff
@@ -85,22 +91,35 @@ class MessagePassing(torch.nn.Module):
         'size_i', 'size_j', 'ptr', 'index', 'dim_size'
     }
 
-    def __init__(self, aggr: Optional[Union[str, List[str]]] = "add",
-                 flow: str = "source_to_target", node_dim: int = -2,
-                 decomposed_layers: int = 1):
+    def __init__(
+        self,
+        aggr: Optional[Union[str, List[str], Aggregation]] = "add",
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:96]: We would need some additional `**aggr_kwargs` passed to the `aggregation_resolver`.

File: torch_geometric/nn/conv/message_passing.py [hunk 3]
```diff
@@ -288,7 +307,7 @@ def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
 
         # Run "fused" message and aggregation (if applicable).
         if (isinstance(edge_index, SparseTensor) and self.fuse
-                and not self.explain and len(self.aggrs) == 0):
+                and not self.explain and (self.aggr in SPMM_AGGRS)):
```
[the remark this answers is not in the document]
Comment (lightaime) [reply torch_geometric/nn/conv/message_passing.py:310]: Not all the `aggrs` support `SparseTensor`. If `message_and_aggregate` is specified, some aggrs such as `std` may throw value errors. Do we want to support message passing with these aggrs when using `SparseTensor`?

File: torch_geometric/nn/conv/message_passing.py [hunk 4]
```diff
@@ -497,13 +507,6 @@ def message_and_aggregate(self, adj_t: SparseTensor) -> Tensor:
         """
         raise NotImplementedError
 
-    def combine(self, inputs: List[Tensor]) -> Tensor:
```
[the remark this answers is not in the document]
Comment (lightaime) [reply torch_geometric/nn/conv/message_passing.py]: Sounds good. Added it back. But the multi aggrs logic is removed. Not sure if we used `combine` anywhere.

File: torch_geometric/nn/conv/message_passing.py [hunk 5]
```diff
@@ -85,22 +87,25 @@ class MessagePassing(torch.nn.Module):
...
+        decomposed_layers: int = 1,
+    ):
 
         super().__init__()
 
-        if aggr is None or isinstance(aggr, str):
-            assert aggr is None or aggr in AGGRS
-            self.aggr: Optional[str] = aggr
-            self.aggrs: List[str] = []
+        if aggr is None or isinstance(aggr, (str, Aggregation)):
```
Comment (lightaime) [inline torch_geometric/nn/conv/message_passing.py:100]: TODO: Make SPMM aggrs such as 'sum' and SumAggregation behave the same

File: torch_geometric/nn/aggr/basic.py [hunk 6]
```diff
@@ -22,7 +22,9 @@ def forward(self, x: Tensor, index: Optional[Tensor] = None, *,
         return self.reduce(x, index, ptr, dim_size, dim, reduce='sum')
 
 
-AddAggregation = SumAggregation  # Alias
+class AddAggregation(SumAggregation):
```
Comment (Padarn) [inline torch_geometric/nn/aggr/basic.py:25]: Do we need `AddAggregation`?

File: torch_geometric/nn/aggr/multi.py [hunk 7]
```diff
@@ -19,7 +20,18 @@ def __init__(self, aggrs: List[Union[Aggregation, str]]):
             raise ValueError(f"'aggrs' of '{self.__class__.__name__}' should "
                              f"not be empty")
 
-        self.aggrs = [aggregation_resolver(aggr) for aggr in aggrs]
+        if aggrs_kwargs is None:
+            aggrs_kwargs = [{}] * len(aggrs)
+        elif len(aggrs) != len(aggrs_kwargs):
+            raise ValueError(f"'aggrs_kwargs' with invalid length passed to "
```
Comment (Padarn) [inline torch_geometric/nn/aggr/multi.py:26]: This error seems like something pretty generic. Do you think there is any value in having a util to generate such an error?

File: torch_geometric/nn/conv/message_passing.py (hunk 5 above)
Comment (Padarn) [reply torch_geometric/nn/conv/message_passing.py:103]: Curious what the gap is today?

File: torch_geometric/nn/conv/message_passing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Padarn) [reply torch_geometric/nn/conv/message_passing.py:299]: Maybe it makes sense to raise the error within the `aggr` itself? Means one less place to update when and if we update the support.

File: torch_geometric/nn/conv/message_passing.py [hunk 8]
```diff
@@ -466,26 +469,20 @@ def explain_message(self, inputs: Tensor, size_i: int) -> Tensor:
         return inputs * edge_mask.view(size)
 
     def aggregate(self, inputs: Tensor, index: Tensor,
-                  ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
-                  aggr: Optional[str] = None) -> Tensor:
```
Comment (Padarn) [inline torch_geometric/nn/conv/message_passing.py:470]: any chance we'll be breaking code by doing this? Seems like a class people may already be building ontop of?

File: torch_geometric/nn/conv/message_passing.py [hunk 9]
```diff
@@ -466,26 +469,20 @@ def explain_message(self, inputs: Tensor, size_i: int) -> Tensor:
...
         """
-        aggr = self.aggr if aggr is None else aggr
-        assert aggr is not None
-        if ptr is not None:
-            ptr = expand_left(ptr, dim=self.node_dim, dims=inputs.dim())
-            return segment_csr(inputs, ptr, reduce=aggr)
-        else:
-            return scatter(inputs, index, dim=self.node_dim, dim_size=dim_size,
-                           reduce=aggr)
+        return self.aggr_module(inputs, index=index, ptr=ptr,
```
Comment (Padarn) [inline torch_geometric/nn/conv/message_passing.py:484]: So much cleaner 💌

File: torch_geometric/nn/aggr/multi.py (hunk 7 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/multi.py:26]: Fixed.

File: torch_geometric/nn/conv/message_passing.py (hunk 8 above)
Comment (rusty1s) [reply torch_geometric/nn/conv/message_passing.py:470]: `aggr` was just introduced in master and now we are removing it. I think we are safe.

File: torch_geometric/nn/aggr/basic.py (hunk 6 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/basic.py]: Removed it.

File: torch_geometric/nn/conv/message_passing.py (hunk 9 above)
Comment (rusty1s) [reply torch_geometric/nn/conv/message_passing.py]: Yes, indeed :)

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
Also fixed jit issues:
* `*` was not allowed as argument in `forward`
* `MultiAggregation` needs to wrap its child aggregation inside `torch.nn.ModuleList`

File: torch_geometric/nn/conv/message_passing.py (hunk 5 above)
Comment (lightaime) [reply torch_geometric/nn/conv/message_passing.py]: The gap is that `self.aggr = str(SumAggregation())` is not the same as 'self.aggr = sum'. Therefore, `self.fuse` is `False` even through `SumAggregation()` should be able to be fused. 
https://github.com/pyg-team/pytorch_geometric/blob/0f7e018aff9f9ca5356bf90b27e3546e26343c37/torch_geometric/nn/conv/message_passing.py#L158

Ways to fix: 
1. Modify the `__repr__` of the fused `Aggregation` functions to return str like `sum`, `mean`, `max`, ...
2. Add their `__repr__` str to `FUSE_AGGRS`. 

Both if them seem to be tricky. Any suggestion?

Comment (rusty1s) [reply torch_geometric/nn/conv/message_passing.py]: Yes, it is tricky. I'm not sure we desperately need to fix this right now, but one option would be to maintain a `repr` variable inside each `Aggregation` module for the reverse mapping. One could even automate that potentially (like a reverse `aggregation_resolver`)

Comment (rusty1s) [reply torch_geometric/nn/conv/message_passing.py]: Added this to the roadmap.

**Pull request merged**